### PR TITLE
Ewb 1691: Add LvFeeder

### PR DESCRIPTION
**Pull request by Lax125**
This will remain a draft at least until https://github.com/zepben/evolve-sdk-jvm/pull/105 is finalized.

### CODE REVIEW

**Comment by ryanjcoleman** [comment]
Task linked: [EWB-1691 Update Python SDK for spec](https://app.clickup.com/t/6929263/EWB-1691)

**Comment by kgreav** [comment]
Again don't merge this one until I'm ready - Clyde is currently working on getting the centralised actions in place for python libs, he should be done soon.

**Comment by charlta** [comment]
Rebased onto `main` to pick up testing changes and resolve conflicts.

File: src/zepben/evolve/testing/test_network_builder.py [hunk 1]
```diff
@@ -299,6 +299,18 @@ def add_feeder(self, head_mrid: str, sequence_number: Optional[int] = None) -> '
         self._create_feeder(self.network.get(head_mrid, ConductingEquipment), sequence_number)
         return self
 
+    def add_lv_feeder(self, head_mrid: str, sequence_number: Optional[int] = None) -> 'TestNetworkBuilder':
```
Comment (charlta) [inline src/zepben/evolve/testing/test_network_builder.py:302]: todo: not running the `AssignToLvFeeders` trace.

File: test/cim/cim_creators.py [hunk 2]
```diff
@@ -1108,6 +1109,20 @@ def create_loop(include_runtime: bool = True):
     )
 
 
+def create_lv_feeder(include_runtime: bool = True):
+    runtime = {
+        "normal_energizing_feeders": lists(builds(Feeder, **create_identified_object(include_runtime)), min_size=1, max_size=2),
+        "current_equipment": lists(builds(Equipment, **create_identified_object(include_runtime)), min_size=1, max_size=2)
```
Comment (charlta) [inline test/cim/cim_creators.py:1115]: todo: This is building `Equipment` which is not a leaf class. I am not sure if/how this is even working. It should be using `sampled_equipment` like the `create_feeder` call.

File: test/cim/cim_creators.py [hunk 3]
```diff
@@ -427,7 +427,7 @@ def create_connectivity_node_container(include_runtime: bool):
 
 def create_equipment(include_runtime: bool):
     runtime = {
-        "current_feeders": lists(builds(Feeder, **create_identified_object(include_runtime)), min_size=1, max_size=2)
+        "current_containers": lists(builds(Feeder, **create_identified_object(include_runtime)), min_size=1, max_size=2)
```
Comment (charlta) [inline test/cim/cim_creators.py:430]: todo: This is only creating a collection of `Feeder` which will hide any left over issues with the change to supporting other equipment containers. It should be updated to use something like `equipment_containers` using `sampled_equipment_container` , but only selecting valid current containers.

File: test/cim/iec61970/infiec61970/feeder/test_lv_feeder.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+#  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+from hypothesis import given
+from hypothesis.strategies import builds, lists
+
+from test.cim.collection_validator import validate_collection_unordered
+from test.cim.iec61970.base.core.test_equipment_container import equipment_container_kwargs, verify_equipment_container_constructor_default, \
+    verify_equipment_container_constructor_kwargs, verify_equipment_container_constructor_args, equipment_container_args
+from zepben.evolve import Feeder, Terminal, Substation, Equipment, LvFeeder
+
+lv_feeder_kwargs = {
```
Comment (charlta) [inline test/cim/iec61970/infiec61970/feeder/test_lv_feeder.py:14]: issue: This test has not been updated to check that `LvFeeder` works, but is still using `Feeder`

File: test/database/sqlite/schema_utils.py [hunk 5]
```diff
@@ -500,6 +500,9 @@ def _add_with_references(filled: T, service: BaseService):
                 it.add_energized_loop(filled)
                 service.add(it)
 
+        if isinstance(filled, LvFeeder):
```
Comment (charlta) [inline test/database/sqlite/schema_utils.py:503]: question: Does this need to add references to the feeder if it is a `LvFeeder`? I am not sure why this is not failing somewhere, perhaps I am missing something, perhaps there is a missing test.

File: test/services/common/translator/base_test_translator.py [hunk 6]
```diff
@@ -58,6 +58,9 @@ def _remove_unsent_references(cim: T):
     if isinstance(cim, ConnectivityNode):
         cim.clear_terminals()
 
+    if isinstance(cim, LvFeeder):
```
Comment (charlta) [inline test/services/common/translator/base_test_translator.py:61]: todo: The check for `Feeder` and `LvFeeder` to call `clear_current_equipment()` can be merged into the check for `EquipmentContainer`

File: test/services/network/test_network_service_comparator.py [hunk 7]
```diff
@@ -923,3 +930,15 @@ def test_compare_loop(self):
             lambda _: Substation(mrid="s1"),
             lambda _: Substation(mrid="s2")
         )
+
+    def test_compare_feeder(self):
```
Comment (charlta) [inline test/services/network/test_network_service_comparator.py:934]: todo: This test overrides the previous test with the same name and prevents it from running. The IDE should be warning you about this.

File: test/services/network/tracing/test_tracing.py [hunk 8]
```diff
@@ -54,7 +55,7 @@ def test_suppliers():
     validate_supplier(tracing.set_direction, SetDirection)
     validate_supplier(tracing.remove_direction, RemoveDirection)
 
-    validate_supplier(tracing.assign_equipment_containers_to_feeders, AssignToFeeders)
+    validate_supplier(tracing.assign_equipment_to_feeders, AssignToFeeders)
```
Comment (charlta) [inline test/services/network/tracing/test_tracing.py:58]: todo: Missing test for `assign_equipment_to_lv_feeders`

**Review by charlta — CHANGES_REQUESTED** [review]
Overall this PR felt rushed.
* The new class this PR is adding is mostly uncovered due to copy/paste issues which are easy to see in the code coverage.
* There were a few instances where the same correction has been made here as was previously covered in the JVM implementation. 

Please take a little longer with your changes to make sure you can strip out the simple/careless things before requesting a review of a PR. I find doing a self review of a PR before adding reviewers can be helpful for this.

File: test/database/sqlite/schema_utils.py (hunk 5 above)
Comment (Lax125) [reply test/database/sqlite/schema_utils.py:500]: `_add_with_references` is only used in `*_services_of`, which are only used in database tests. `filled` never has runtime fields (e.g. `Feeder().current_equipment`).

File: src/zepben/evolve/model/cim/iec61970/base/core/equipment.py [hunk 9]
```diff
@@ -275,10 +296,10 @@ def clear_usage_points(self) -> Equipment:
 
     def get_operational_restriction(self, mrid: str) -> OperationalRestriction:
         """
-        Get the `zepben.evolve.cim.iec61968.operations.operational_restriction.OperationalRestriction` for this `Equipment` identified by `mrid`
+        Get the `OperationalRestriction` for this `Equipment` identified by `mrid`
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/core/equipment.py:300]: I've removed it this file, I'll make a different PR to remove them for the rest. Sorry about the noise, the only other out-of-scope changes in this PR is removing unused imports from my previous code.

[the remark this answers is not in the document]
Comment (charlta) [reply src/zepben/evolve/model/cim/iec61970/base/core/equipment.py:300]: If you create a branch and remove these for all files now I can merge it and you can rebase to remove the noise.

**Review by charlta — CHANGES_REQUESTED** [review]

**Review by charlta — APPROVED** [review]

**Pull request merged**